sopel_modules/github/webhook.py

```diff
@@ -107,13 +112,73 @@ def get_targets(repo):
+
+
+def abort_request(status_code=400, response_message=None):
+    if sopel_instance.config.github.debug_mode:
+        LOGGER.warning('`debug_mode = True`; allowing unverified request...')
+        return None
+    return bottle.abort(status_code, response_message)
+
+
+def verify_request():
+    request_headers = bottle.request.headers
+    request_body = bottle.request.body.read()
+
+    if not request_headers.get('X-Hub-Signature'):
+        msg = 'Request is missing a hash signature.'
+        LOGGER.error(msg)
+        debug_log_request(request_headers, request_body)
+        return abort_request(401, msg)  # 401 Unauthorized; missing required header
+
+    digest_name, payload_signature = request_headers.get('X-Hub-Signature').split('=')
+    # Currently, GitHub only uses 'sha1'; log a warning if a different digest is
+    # specified by the server. GitHub may have started using a new digest and
+    # this should be confirmed.
+    if digest_name != 'sha1':
+        LOGGER.warning('Unexpected signature digest: {}'.format(digest_name))
+        debug_log_request(request_headers, request_body)
+
+    try:
+        digest_mod = getattr(hashlib, digest_name)
+    except AttributeError:
+        # The previous digest check does not require a 'sha1' digest, but simply
+        # warns when an unexpected digest is specified. The function will
+        # attempt to find the digest specified in the signature, but if it is
+        # not currently supported by Python's `hashlib`, an error will be logged
+        # and returned.
+        msg = 'Unsupported signature digest: {}'.format(digest_name)
+        LOGGER.error(msg)
+        debug_log_request(request_headers, request_body)
+        return abort_request(501, msg)  # 501 Not Implemented; server does not support the functionality required to fulfill the request
+
+    secret = sopel_instance.config.github.webhook_secret
+    hash_ = hmac.new(secret.encode('utf-8') if secret else None, msg=request_body, digestmod=digest_mod)
+    expected_signature = hash_.hexdigest()
+    if payload_signature != expected_signature:
+        msg = 'Request signature mismatch.'
+        LOGGER.error(msg)
+        debug_log_request(request_headers, request_body)
+        return abort_request(403, msg)  # 403 Forbidden; server understood the request but refuses to authorize it
+
+
 @bottle.get("/webhook")
 def show_hook_info():
-    return 'Listening for webhook connections!'
+    if sopel_instance.config.github.debug_mode:
+        return 'Listening for webhook connections!'
+    # bottle.abort() == raise HTTPError(); manually raising HTTPError allows passing extra headers
+    raise bottle.HTTPError(405, Allow='POST')  # 405 Method Not Allowed; this route is only useful for testing.
 
 
 @bottle.post("/webhook")
 def webhook():
+    if sopel_instance.config.github.webhook_secret:
+        verify_request()  # Bottle.py will automatically abort this webhook for `abort`'d requests.
```

Comment:
Might want to update this comment, since the path to Bottle's `abort` method is pretty deep.

```suggestion
        verify_request()  # function will automatically abort this webhook if verification fails
```